File: abstraction.py
# Abstraction hides the implementation details and exposes only relevant features to the user.

from abc import ABC, abstractmethod


class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass


class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3.14 * self.radius * self.radius

    def perimeter(self):
        return 2 * 3.14 * self.radius


class Rectangle(Shape):
    def __init__(self, length, breadth):
        self.length = length
        self.breadth = breadth

    def area(self):
        return self.length * self.breadth

    def perimeter(self):
        return 2 * (self.length + self.breadth)


circle = Circle(7)

print("Area of circle:", circle.area())
print("Perimeter of circle:", circle.perimeter())

rectangle = Rectangle(5, 10)

print("Area of rectangle:", rectangle.area())
print("Perimeter of rectangle:", rectangle.perimeter())

# In the above example, the Shape class is an abstract class that defines two abstract methods: area and perimeter. The Circle and Rectangle classes inherit from the Shape class and implement these abstract methods. The abstract methods enforce that any subclass of the Shape class must implement these methods. This allows us to define a common interface for different shapes while hiding the implementation details of each shape.

File: abstraction.cpp
// Abstraction hides the implementation details and exposes only relevant features to the user.

#include <bits/stdc++.h>
using namespace std;

class Shape
{
public:
    virtual void draw() = 0;
};

class Circle : public Shape
{
public:
    void draw() override
    {
        cout << "Drawing Circle" << endl;
    }
};

class Rectangle : public Shape
{
public:
    void draw() override
    {
        cout << "Drawing Rectangle" << endl;
    }
};

int main()
{
    Circle c;
    Rectangle r;
    c.draw();
    r.draw();
    return 0;
}

// In the above code, we have a base class Shape with a pure virtual function draw(). The Circle and Rectangle classes inherit from the Shape class and implement the draw() method. The draw() method is an abstract method in the Shape class, and it is overridden in the Circle and Rectangle classes. When we create objects of the Circle and Rectangle classes, we can call the draw() method to draw the respective shapes. This is an example of abstraction, where the implementation details of drawing the shapes are hidden from the user, and only the relevant feature (draw) is exposed.